### ISSUE TITLE
[Issue]: Invalid kernel launch grid dimensions error does not clear hip device state

### ISSUE THREAD

**Issue by giuseppegrossi** [ROCmSoftwarePlatform/triton#943]
### Problem Description

When launching a triton kernel with a launch grid dimension size that exceeds hardware capacity, triton raises a runtime error as expected. What's interesting is that if you handle that runtime error any follow up gpu operation, like a torch op, will cause a subsequent crash because it has a lingering HIP error unless it has been cleared. So this later hip related crash can happen unexpectedly and the error may not be self evident since this can happen much later in a different place. I think the triton kernel error may need to clear the device state via a hipGetErrorMessage() call .

### Operating System

Ubuntu 22.04.4 LTS (Jammy Jellyfish)

### CPU

AMD EPYC 9654 96-Core Processor

### GPU

AMD Instinct MI300X

### ROCm Version

7.2

### Steps to Reproduce

```python
"""
Reproducer: Launch a Triton kernel with invalid grid sizes, catch the error,
then perform a basic PyTorch op to check for lingering HIP errors.
"""
import torch
import triton
import triton.language as tl
import numpy as np


@triton.jit
def dummy_kernel(output_ptr, n_elements, BLOCK_SIZE: tl.constexpr):
    pid = tl.program_id(0)
    offsets = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
    mask = offsets < n_elements
    tl.store(output_ptr + offsets, 1.0, mask=mask)


def main():
    torch.cuda.synchronize()
    print("=== Step 1: Launch Triton kernel with invalid grid size ===")

    output = torch.zeros(1024, device="cuda")

    # Use an absurdly large grid to trigger a launch failure
    bad_grid = (2, 130000, 1)  # way too large

    try:
        dummy_kernel[bad_grid](output, output.numel(), BLOCK_SIZE=1024)
        print("ERROR: kernel launch did not fail (unexpected)")
    except Exception as e:
        print(f"Caught expected error: {type(e).__name__}: {e}")

    print("\n=== Step 2: Attempt basic numpy operation ===")
    try:
        a = np.random.randn(256, 256)
        b = np.random.randn(256, 256)
        c = a @ b
        print(f"MatMul succeeded, c.sum() = {c.sum().item():.4f}")
    except RuntimeError as e:
        print(f"NUMPY ERROR detected: {type(e).__name__}: {e}")

    print("\n=== Step 3: Attempt basic PyTorch operation ===")
    try:
        a = torch.randn(256, 256, device="cuda")
        b = torch.randn(256, 256, device="cuda")
        c = a @ b
        print(f"MatMul succeeded, c.sum() = {c.sum().item():.4f}")
    except RuntimeError as e:
        print(f"LINGERING HIP ERROR detected: {type(e).__name__}: {e}")

    
    print("\n=== Step 4: Attempt basic PyTorch operation folowup ===")
    try:
        a = torch.randn(256, 256, device="cuda")
        b = torch.randn(256, 256, device="cuda")
        c = a @ b
        print(f"MatMul succeeded, c.sum() = {c.sum().item():.4f}")
    except RuntimeError as e:
        print(f"LINGERING HIP ERROR detected: {type(e).__name__}: {e}")

if __name__ == "__main__":
    main()

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

_No response_

### Additional Information

You can use the latest vllm image, vllm/vllm-openai-rocm:latest to have a reproducible environment.

Explanation for the reproducer script is as follows:
Step 1: launched a bad kernel grid and handles runtime error
Step 2: Does a numpy operation to show cpu ops work fine
Step 3: Does a torch operation and we get a HIP invalid argument error (this is the problem)
Step 4: Redoes the same thing as step 3 but it works since error has been encountered and cleared already

When the triton kernel crashes and fails to launch and gets handle by user code should it not clear the faulty device state at the same time too?
